### PR TITLE
Factorized the code for allow_access and allow_publish

### LINKED ISSUE #100: Security concerns
I am currently trying to implement #99 and noticed that anybody can access any package as long as they are authenticated through Gitlab. (https://github.com/bufferoverflow/verdaccio-gitlab/blob/v2.2.0/src/gitlab.js#L158)

From what I understand, to check whether the user is allowed to access a package or not, we only check if its user is set.

There is no verification made to ensure the user is accessing a scope he owns or that he has the proper access rights.

I am currently working on #99 so I could tackle this issue as well, I think. We just have to properly discuss the approach to solve it.

Here is what I planned to do to solve #99:
- Add a configuration option `allowAnyScope`. This tells the plugin that accessing a non-owned scope might be allowed.
- If `allowAnyScope` is true, we don't fail when `packagePermit` is false. Instead, we retrieve `_package.publish` and check if its value is within `user.real_groups`.
- If the value is found, we allow the publishing.

Doing this was easy but I didn't expect to be able to see/access my published package. That's how I found the security concern.

### Proposed solution:

I think the behavior for `allow_access` should follow the same pattern as `allow_publish`:
If the user doesn't have access to the group/project, it cannot access the packages within that scope.

### PR DESCRIPTION
**Pull request by Telokis**
Closes #100 

This PR aims to fix the security issues the current `allow_access` implementation has.  
Anybody can access any package as long as he is authenticated.  
Instead of just fixing this issue, the new branch completely merges the behavior of `allow_access` and `allow_publish` by introducing a generic `allow_action` function that is used by both of them.  
The new behavior allows users to specify `$authenticated` and `$all`/`$anonymous` for both `publish` and `access`.  
It also introduces a new meta-group: `$owned-group` (This name could change before the release). This meta-group represents a permission where the user can only perform the action on a package where he has `publishLevel` permission in Gitlab. This is the curent behavior for `publish`.

### CODE REVIEW

**Comment by Telokis** [comment]
I updated the tests and the README (I took the liberty to incorporate and adapt the changes suggested by @slhck in #104 ).

File: README.md [hunk 1]
```diff
@@ -52,13 +52,13 @@ packages:
   '@*/*':
     # scoped packages
     access: $all
-    publish: $authenticated
+    publish: $owned-group
```
Comment (bufferoverflow) [inline README.md]: This should e.g. be `$maintainer `as described here https://github.com/bufferoverflow/verdaccio-gitlab#configuration-options.

File: README.md [hunk 2]
```diff
@@ -93,40 +93,53 @@ are marked with `gitlab: true` as in the configuration sample above. If you
...
+handles the following internal groups to determine permissions:
 
-- authenticated users are able to access all packages
-- unauthenticated users will be able to access packages marked with either
-  `$all` or `$anonymous` access levels at the package group definition
+- `$all` or `$anonymous`: Any user can perform the action on the packages. No
+condition at all.
+- `$authenticated`: Any successfully authenticated user can perform the action
+on the packages.
+- `$owned-group`: A user can perform the action on a package if
```
Comment (bufferoverflow) [inline README.md]: We need to take care on the access level and I think there is no need for `$owned-group` as we have access level for this.

File: package.json [hunk 3]
```diff
@@ -6,7 +6,7 @@
   },
   "scripts": {
     "type-check": "tsc --noEmit",
-    "commitmsg": "commitlint -e $GIT_PARAMS",
+    "commitmsg": "commitlint -q -e $GIT_PARAMS",
```
Comment (bufferoverflow) [inline package.json]: Why is this needed?

File: src/gitlab.ts [hunk 4]
```diff
@@ -37,6 +37,7 @@ const BUILTIN_ACCESS_LEVEL_ANONYMOUS = ['$anonymous', '$all'];
 
 // Level to apply on 'allow_access' calls when a package definition does not define one
 const DEFAULT_ALLOW_ACCESS_LEVEL = ['$all'];
+const DEFAULT_ALLOW_PUBLISH_LEVEL = ['$owned-group'];
```
Comment (bufferoverflow) [inline src/gitlab.ts]: should be `$owner`

File: README.md (hunk 1 above)
Comment (Telokis) [reply README.md]: I thought about doing it this way but the issue is that I would need to retrieve all user's projets/groups. The current way is to retrieve only the groups/projets where the user is at least `auth.gitlab.publish` meaning that I would have to change that as well.

File: README.md (hunk 2 above)
Comment (Telokis) [reply README.md]: See my answer above, I would need to alter the way the extension retrieve the user's data.

File: package.json (hunk 3 above)
Comment (Telokis) [reply package.json]: I kept getting a `formatted.join` when executing `commitlint` commands. Absolutely no idea why and I didn't feel like investigating the library.

File: src/gitlab.ts (hunk 4 above)
Comment (Telokis) [reply src/gitlab.ts]: If I were to do it the way you suggest, I think the default value should be `$maintainer` since it currently is the default value in the code. (https://github.com/bufferoverflow/verdaccio-gitlab/blob/2fa6944cc3f43c31f32127a61bb4e09b22504461/src/gitlab.ts#L68)

File: package.json (hunk 3 above)
Comment (Telokis) [reply package.json]: But I agree that I shouldn't quiet it. It totally disables the error output which makes fixinx lint errors annoying.

File: README.md (hunk 1 above)
Comment (Telokis) [reply README.md]: Originally, when working on this PR, I decided to use `$owned-group` as a way to represent the current behavior that the plugin is using for `publish` authorization.
The naming may not be ideal (though I have nothing better in mind, yet) but it's there for the user to have a way to say "Only do this action if you have at least `auth.gitlab.publish` permission on the associated group/project".

I think adding support for `$maintainer`, `$owner`, ... beyond the `auth.gitlab.publish` should be another PR with a discussion regarding the design itself.

I'm not sure you remember my other issue but I made this PR because I plan to do #99 right after it's merged.

File: package.json (hunk 3 above)
Comment (Telokis) [reply package.json]: Regarding this, I also noticed a deprecation warning:
```
$ yarn run commitmsg
yarn run v1.19.0
$ commitlint -e $GIT_PARAMS
Using environment variable syntax ($GIT_PARAMS) in -e |--edit is deprecated. Use '{-E|--env} HUSKY_GIT_PARAMS instead'
C:\Users\Telokis\Documents\git\verdaccio-gitlab\node_modules\@commitlint\cli\lib\cli.js:119
        throw err;
        ^

Error: Received $GIT_PARAMS as value for -e | --edit, but GIT_PARAMS or HUSKY_GIT_PARAMS are not available globally.
    at getEditValue (C:\Users\Telokis\Documents\git\verdaccio-gitlab\node_modules\@commitlint\cli\lib\cli.js:307:9)
    at normalizeFlags (C:\Users\Telokis\Documents\git\verdaccio-gitlab\node_modules\@commitlint\cli\lib\cli.js:268:15)
    at C:\Users\Telokis\Documents\git\verdaccio-gitlab\node_modules\@commitlint\cli\lib\cli.js:126:11
    at new Promise (<anonymous>)
    at main (C:\Users\Telokis\Documents\git\verdaccio-gitlab\node_modules\@commitlint\cli\lib\cli.js:123:9)
    at Object.<anonymous> (C:\Users\Telokis\Documents\git\verdaccio-gitlab\node_modules\@commitlint\cli\lib\cli.js:115:1)
    at Module._compile (internal/modules/cjs/loader.js:759:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:770:10)
    at Module.load (internal/modules/cjs/loader.js:628:32)
    at Function.Module._load (internal/modules/cjs/loader.js:555:12)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I can easily fix it by removing `$GIT_PARAMS` from the arguments.

Comment (Telokis) [reply package.json]: I pushed this fix and reverted the `-q`

**Comment by Telokis** [comment]
By the way, I didn't mention it but I think this should be a major release, it will break all existing configurations relying on `$authenticated`.

File: src/gitlab.ts (hunk 4 above)
Comment (bufferoverflow) [reply src/gitlab.ts]: yep, this makes sense

File: README.md (hunk 1 above)
Comment (Telokis) [reply README.md]: @bufferoverflow What should I do regarding this comment?

Comment (bufferoverflow) [reply README.md]: `$maintainer` is the current default as figured out [here](https://github.com/bufferoverflow/verdaccio-gitlab/pull/101/files#r355184353), so we do not need `$owned-group`

Comment (Telokis) [reply README.md]: Those two points are independent, I specifically said `If I were to do it the way you suggest` at the beginning of my message because it is not the way I designed it to work in this PR.

Sorry if I'm not clear enough, I'll try to better explain:

The main aim of this PR is to provide a slightly more intuitive/explicit configuration.  
The current behavior of the plugin is to do something very specific for `publish` and `access` authorizations. There is no way to customize the behavior: I can't allow `$anonymous` to `publish` and I can't enforce the current `publish` behavior for `access` authorization.  

This PR attempts to fix this issue with two reworks at once:
- Unify `allow_publish` and `allow_access` under a `allow_action` with specific arguments and defaults.
- Explicitly handle the different use-cases the user may be willing to use (for example `$anonymous` which allows anyone to perform the action or `$authenticated` that allow any authenticated user to perform the action)

Handling the first point was fairly easy but the second point requires more explicit configuration from the user. He can't simply use anything as the value of `publish` or `access`. He needs to state the exact behavior the module will use.

Finally, this stricter/more explicit handling of permissions created the need for some group to represent the old `publish` behavior. That's what `$owned-group` solves even though the name is not ideal.


I hope this explanation is clear.

Comment (Telokis) [reply README.md]: @bufferoverflow Any feedback regarding my explanation?

Comment (bufferoverflow) [reply README.md]: I currently prefer separate handling instead of a unified function and I think caching should be enabled for this, otherwise you DDoS the gitlab server with each npm install.

Comment (Telokis) [reply README.md]: > I currently prefer separate handling instead of a unified function

Do you have any reason for this?

> I think caching should be enabled for this, otherwise you DDoS the gitlab server with each npm install.

Where do I call the gitlab server? I copy/pasted the previous code and unified it

Comment (bufferoverflow) [reply README.md]: caching is done based on the setting within publish level, this is used within your code as well, so it only works if access and publish level is the same when cache is enabled.

Comment (Telokis) [reply README.md]: Do you mean that the `authenticate` function is called often?  
Even if that's the case, the user is cached with `_setCachedUserGroups` so it won't call the Gitlab API more than once per user.  
`access` and `publish` permissions both rely on `publishLevel` when `"$owned-group"` is specified.

Comment (bufferoverflow) [reply README.md]: https://github.com/bufferoverflow/verdaccio-gitlab/blob/master/src/authcache.ts#L50 does not have the access level, so if we add functionality for different access levels for access, people should be able to define the level of their choice for access and publish.

Comment (Telokis) [reply README.md]: Yes, for now they will both use the `publishLevel` value without distinction.  
The proper way would be to be able to specify `$maintainer`, `$owner` and so on for both `publish` and `access` but this requires much more work. (as mentioned here: https://github.com/bufferoverflow/verdaccio-gitlab/pull/101#discussion_r355183736)

Comment (Telokis) [reply README.md]: Do you have any other feedback or is this PR ok as-is?

Comment (bufferoverflow) [reply README.md]: as said, **if we add functionality for different access levels, people should be able to define the level of their choice for access and publish.**

Comment (Telokis) [reply README.md]: Yes but this is not the goal of this PR. Its only goal is to increase security.

**Comment by bufferoverflow** [comment]
> as said, if we add functionality for different access levels, people should be able to define the level of their choice for access and publish.

https://github.com/bufferoverflow/verdaccio-gitlab/pull/101#discussion_r365688342
